### PR TITLE
Migrate applicant

### PR DESCRIPTION
**Pull request by rebeccabrennan**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/FPLA-382

### Change description ###
OLD

`
{"caseName": "test_applicant_case", "applicant": {"name": "test_name", "email": null, "mobile": null, "address": {"County": null, "Country": null, "PostCode": null, "PostTown": null, "AddressLine1": null, "AddressLine2": null, "AddressLine3": null}, "jobTitle": null, "pbaNumber": null, "telephone": null, "personToContact": null}, "solicitor": {"dx": null, "name": null, "email": null, "mobile": null, "reference": null, "telephone": null}, "caseLocalAuthority": "HN"}
`

AFTER MIGRATION

`
{"caseName": "test_applicant_case", "solicitor": {"dx": null, "name": null, "email": null, "mobile": null, "reference": null, "telephone": null}, "applicants": [{"id": "84d975d5-d245-4600-9b0a-ed0203f8bff4", "value": {"party": {"address": {}, "partyId": "045478a6-43e1-4853-a44c-489f77794475", "partyType": "ORGANISATION", "organisationName": "test_name"}, "leadApplicantIndicator": "Yes"}}], "caseLocalAuthority": "HN"}
`

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationService.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        telephoneNumberBuilder.contactDirection(defaultIfBlank(or.getPersonToContact(), null));
+        TelephoneNumber telephoneNumber = telephoneNumberBuilder.build();
+
+        MobileNumber.MobileNumberBuilder mobileNumberBuilder = MobileNumber.builder();
+        mobileNumberBuilder.telephoneNumber(defaultIfBlank(or.getMobile(), null));
+        MobileNumber mobileNumber = mobileNumberBuilder.build();
+
+        ApplicantParty.ApplicantPartyBuilder partyBuilder = ApplicantParty.builder();
+        partyBuilder.partyId(UUID.randomUUID().toString());
+        partyBuilder.partyType("Individual");
```
Comment (geek-soft) [inline src/main/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationService.java]: Ideally if we used ENUM for that.

File: src/main/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationService.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        if (!isEmpty(or.getName())) {
+            partyBuilder.name(defaultIfBlank(or.getName().split("\\s+")[0], null));
```
Comment (geek-soft) [inline src/main/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationService.java]: That looks like we are disregarding surname. Looks bad 👀 

What is more split logic should be tested as a unit (separate method and separate unit tests)

File: src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+    private final String PERSONTOCONTACT = "Person to contact";
+    private final String PARTYTYPE = "ApplicantParty type";
+    private final String LEADAPPLICANTINDICATOR = "Yes";
+    private final String PARTYID = UUID.randomUUID().toString();
+    private final String PBANUMBER= "PBA123456";
+
+    private final ApplicantDataMigrationService service = new ApplicantDataMigrationService();
+
+    @Test
+    void whenOldStructureDoesNotExistAcceptsShouldReturnFalse() {
```
Comment (geek-soft) [inline src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java]: In ideal world you could use concept of nested test. I am not saying you have to, that code will never be merged it should just work but just wanted to share good practice with you.

Here is doc: https://junit.org/junit5/docs/current/user-guide/#writing-tests-nested

File: src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+        assertThat(service.accepts().test(caseDetails)).isEqualTo(false);
+    }
+    @Test
+    void whenDataIsNullAcceptsShouldReturnFalse() {
+        CaseDetails caseDetails = CaseDetails.builder()
+            .id(1111L)
+            .data(null)
+            .build();
+
+        assertThat(service.accepts().test(caseDetails)).isEqualTo(false);
```
Comment (geek-soft) [inline src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java]: Why you haven't added test case for accepted case? If logic was completely broken and always returned false your tests would not know about it. Please add it so there are no surprises.

File: src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+    }
+
+    @Test
+    void whenOldApplicantStructureIsMigratedShouldReturnNewAplicantStructure() {
+        Map<String, Object> data = new HashMap<>();
+
+        OldApplicant applicant = getOldApplicant();
+
+        Applicant newApplicant = newApplicantBuilder();
```
Comment (geek-soft) [inline src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java]: Why is that declared so early if it is used in last line?
Good rule says keep declaration as close as possible to the place it is used for the first time.

File: src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+                    .telephoneNumber(null)
+                    .build())
+                .jobTitle(null)
+                .pbaNumber(null)
+                .build())
+            .leadApplicantIndicator(null)
+            .build()));
+    }
+
+    private Applicant newApplicantBuilder() {
```
Comment (geek-soft) [inline src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java]: Wired name, sort of suggests that object builder is returned but no, surprise, complete instance is returned instead.

File: src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+                    .build())
+                .jobTitle(JOBTITLE)
+                .pbaNumber(PBANUMBER)
+                .build())
+            .leadApplicantIndicator(LEADAPPLICANTINDICATOR)
+            .build();
+        return newApplicant;
+    }
+
+    private OldApplicant getOldApplicant() {
```
Comment (geek-soft) [inline src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java]: Get method does not quite fit here for me. In my mind get gives you something that is already present but here creation process happens. Bit misleading for me.

File: src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+
+class ApplicantDataMigrationServiceTest {
+    private final String APPLICANT = "Applicant name";
+    private final String EMAIL = "email@email.com";
+    private final String MOBILE = "07778987656";
+    private final String COUNTY = "County";
+    private final String COUNTRY = "Country";
+    private final String POSTCODE = "BT17NT";
+    private final String POSTTOWN = "Ballymena";
+    private final String ADDRESSLINE1 ="ADDRESSLINE1";
```
Comment (geek-soft) [inline src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java]: Why these names are without any separator? They are just hard to read. Use `_` as a separator next time.

File: src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+        service.migrate(caseDetails);
+
+
+        List<Map<String, Object>> valueInApplicant = (List<Map<String, Object>>) caseDetails.getData().get("applicants");
+
+        assertThat(valueInApplicant.get(0).get("value").equals(newApplicant));
+    }
+
+    @Test
+    void whenPartiallyFilledInOldApplicantStructureIsMigratedShouldReturnNewListStructureWithNullFields() {
```
Comment (geek-soft) [inline src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java]: None of the tests make sure `applicant` is set to null which is an important part of migration.

File: src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+import uk.gov.hmcts.reform.domain.common.TelephoneNumber;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class ApplicantDataMigrationServiceTest {
```
Comment (geek-soft) [inline src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java]: There is no test for an empty `applicant` object that I guess can happen. Please add.

File: src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+                    .build())
+                .jobTitle(null)
+                .pbaNumber(null)
+                .build())
+                .leadApplicantIndicator(null)
+            .build()));
+    }
+
+    @Test
+    void whenPartiallyFilledInOldApplicantAddressIsMigratedShouldReturnNewListStructureWithNullFields() {
```
Comment (geek-soft) [inline src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java]: I am not quite sure why we have separate case for postcode and county populated? Is that common practice for cases to have only the two populated?

File: src/main/java/uk/gov/hmcts/reform/domain/OldApplicant.java [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+        private final String mobile;
+        private final Address address;
+        private final String jobTitle;
+        private final String telephone;
+        private final String personToContact;
+        private final String pbaNumber;
+
+        @JsonCreator
+        public OldApplicant( @JsonProperty("name") final String name,
+                             @JsonProperty("Email") final String email,
```
[the remark this answers is not in the document]
Comment (rebeccabrennan) [reply src/main/java/uk/gov/hmcts/reform/domain/OldApplicant.java]: 👍

File: src/main/java/uk/gov/hmcts/reform/domain/common/ApplicantParty.java [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+import lombok.Builder;
+import lombok.Data;
+
+@Data
+@Builder
+@AllArgsConstructor
+public class ApplicantParty {
+    private final String partyId;
+    private final String partyType;
+    private final String name;
```
[the remark this answers is not in the document]
Comment (rebeccabrennan) [reply src/main/java/uk/gov/hmcts/reform/domain/common/ApplicantParty.java]: in the new structure i was given it only seems to be name, who would I ask about this @swalker125

File: src/main/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationService.java [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+import java.util.UUID;
+import java.util.function.Predicate;
+
+import static org.apache.commons.lang3.StringUtils.defaultIfBlank;
+import static org.springframework.util.ObjectUtils.isEmpty;
+
+@Slf4j
+@Service
+public class ApplicantDataMigrationService implements DataMigrationService {
+    private ObjectMapper objectMapper = new ObjectMapper();
```
[the remark this answers is not in the document]
Comment (rebeccabrennan) [reply src/main/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationService.java]: 👍

[the remark this answers is not in the document]
Comment (rebeccabrennan) [reply src/main/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationService.java]: 👍

File: src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java (hunk 3 above)
Comment (rebeccabrennan) [reply src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java]: 👍

File: src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java (hunk 5 above)
Comment (rebeccabrennan) [reply src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java]: 👍

File: src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java (hunk 6 above)
Comment (rebeccabrennan) [reply src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java]: 👍

File: src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java (hunk 7 above)
Comment (rebeccabrennan) [reply src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java]: 👍

File: src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java (hunk 8 above)
Comment (rebeccabrennan) [reply src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java]: 👍

File: src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java (hunk 9 above)
Comment (rebeccabrennan) [reply src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java]: 👍

File: src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java (hunk 10 above)
Comment (rebeccabrennan) [reply src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java]: @swalker125 what way would this be tested?

File: src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java (hunk 11 above)
Comment (rebeccabrennan) [reply src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java]: It is to test partial completion of builders like address that the user is still allowed to submit.

File: src/main/java/uk/gov/hmcts/reform/domain/common/ApplicantParty.java [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+    private final Address address;
+    private final EmailAddress emailAddress;
```
[the remark this answers is not in the document]
Comment (rebeccabrennan) [reply src/main/java/uk/gov/hmcts/reform/domain/common/ApplicantParty.java]: so we had problem when just calling it email as there was already email declared so I had to chenge to email address, not sure what way to resolve this.

[the remark this answers is not in the document]
Comment (rebeccabrennan) [reply src/main/java/uk/gov/hmcts/reform/domain/common/ApplicantParty.java]: @swalker125

File: src/main/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationService.java (hunk 2 above)
Comment (rebeccabrennan) [reply src/main/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationService.java]: I dont know, the schema just gives name and the name of a local authroity so im not sure if first name and last name not sure who to ask this to, @swalker125

File: src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java (hunk 4 above)
Comment (rebeccabrennan) [reply src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java]: @swalker125 how would you test this?

File: src/main/java/uk/gov/hmcts/reform/domain/common/ApplicantParty.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (rebeccabrennan) [reply src/main/java/uk/gov/hmcts/reform/domain/common/ApplicantParty.java]: 👍

File: src/main/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationService.java (hunk 2 above)
Comment (rebeccabrennan) [reply src/main/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationService.java]: Dont think i need this as it is just name in my schema? Unless it is actually first and last name although i am not sure need to talk to @geek-soft and @swalker125

File: src/main/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationService.java (hunk 1 above)
Comment (rebeccabrennan) [reply src/main/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationService.java]: 👍

File: src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java (hunk 10 above)
Comment (rebeccabrennan) [reply src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java]: 👍

File: src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java (hunk 4 above)
Comment (rebeccabrennan) [reply src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java]: 👍

File: src/main/java/uk/gov/hmcts/reform/domain/common/ApplicantParty.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (rebeccabrennan) [reply src/main/java/uk/gov/hmcts/reform/domain/common/ApplicantParty.java]: 👍  changed to organisationName

File: src/main/java/uk/gov/hmcts/reform/domain/common/ApplicantParty.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (swalker125) [reply src/main/java/uk/gov/hmcts/reform/domain/common/ApplicantParty.java]: changed

File: src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java (hunk 8 above)
Comment (swalker125) [reply src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java]: made change

File: src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java [hunk 16]
```diff
@@ -0,0 +1,210 @@
...
+
+        CaseData migratedData = service.migrate(data);
+
+        ApplicantParty actualValue = migratedData.getApplicants().get(0).getValue().getParty();
+        ApplicantParty expectedValue = createNewApplicant().getParty();
+
+        assertThat(migratedData.getApplicant()).isNull();
+
+        assertThat(actualValue.getPartyId()).isNotNull();
+        assertThat(actualValue.getPartyType()).isNull();
```
Comment (geek-soft) [inline src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java]: No party type is checked / set. We should check for exact value as we know what it should be.

REQ

File: src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java [hunk 17]
```diff
@@ -0,0 +1,210 @@
...
+
+        data.put("applicant", applicant);
+
+        CaseData migratedData = service.migrate(data);
+
+        ApplicantParty actualValue = migratedData.getApplicants().get(0).getValue().getParty();
+
+        assertThat(migratedData.getApplicant()).isNull();
+        assertThat(actualValue.getPartyId()).isNotNull();
+        assertThat(actualValue.getPartyType()).isNull();
```
Comment (geek-soft) [inline src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java]: In that case party type should not be null either.

REQ

File: src/main/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationService.java [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+        partyBuilder.mobileNumber(mobileNumber);
+        partyBuilder.jobTitle(defaultIfBlank(oa.getJobTitle(), null));
+        partyBuilder.pbaNumber(defaultIfBlank(oa.getPbaNumber(), null));
+        ApplicantParty party = partyBuilder.build();
+
+        CollectionEntry<Applicant> applicantEntry = CollectionEntry.<Applicant>builder()
+            .id(UUID.randomUUID().toString())
+            .value(Applicant.builder()
+                .party(party)
+                .build())
```
[the remark this answers is not in the document]
Comment (geek-soft) [reply src/main/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationService.java]: I'd set it to `No` for every applicant.

[the remark this answers is not in the document]
Comment (swalker125) [reply src/main/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationService.java]: case service sets here 

   return Element.<Applicant>builder()
                        .id(element.getId())
                        .value(applicantBuilder.leadApplicantIndicator("Yes").build())
                        .build();

[the remark this answers is not in the document]
Comment (swalker125) [reply src/main/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationService.java]: have updated migration to match

File: src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java (hunk 17 above)
Comment (swalker125) [reply src/test/java/uk/gov/hmcts/reform/migration/service/ApplicantDataMigrationServiceTest.java]: added assertions around partyType = ORGANISATION

File: build.gradle [hunk 19]
```diff
@@ -34,11 +34,15 @@ def versions = [
 dependencies {
   annotationProcessor group: 'org.projectlombok', name: 'lombok', version: versions.lombok
   compileOnly group: 'org.projectlombok', name: 'lombok', version: versions.lombok
-
+
```
[the remark this answers is not in the document]
Comment (swalker125) [reply build.gradle]: not fussed about this

File: src/main/java/uk/gov/hmcts/reform/domain/common/ApplicantParty.java [hunk 20]
```diff
@@ -0,0 +1,20 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+
+@Data
+@Builder
+@AllArgsConstructor
+public class ApplicantParty {
+    private final String partyId;
+    private final String partyType;
```
[the remark this answers is not in the document]
Comment (swalker125) [reply src/main/java/uk/gov/hmcts/reform/domain/common/ApplicantParty.java]: This should go into the base case migration starter if it were to go anywhere

**Comment by swalker125** [comment]
OLD

`
{"caseName": "test_applicant_case", "applicant": {"name": "test_name", "email": null, "mobile": null, "address": {"County": null, "Country": null, "PostCode": null, "PostTown": null, "AddressLine1": null, "AddressLine2": null, "AddressLine3": null}, "jobTitle": null, "pbaNumber": null, "telephone": null, "personToContact": null}, "solicitor": {"dx": null, "name": null, "email": null, "mobile": null, "reference": null, "telephone": null}, "caseLocalAuthority": "HN"}
`

AFTER MIGRATION

`
{"caseName": "test_applicant_case", "solicitor": {"dx": null, "name": null, "email": null, "mobile": null, "reference": null, "telephone": null}, "applicants": [{"id": "84d975d5-d245-4600-9b0a-ed0203f8bff4", "value": {"party": {"address": {}, "partyId": "045478a6-43e1-4853-a44c-489f77794475", "partyType": "ORGANISATION", "organisationName": "test_name"}, "leadApplicantIndicator": "Yes"}}], "caseLocalAuthority": "HN"}
`

**Pull request closed without merging**